### PR TITLE
CYB-367 - Window scroll in Gallery is making it hard to navigate

### PR DESCRIPTION
**Pull request by nicolaasmatthijs**
https://jira.ets.berkeley.edu/jira/browse/CYB-367

This PR does a number of different things:

- Prevent the full gallery list from being reloaded on every page view
- Implement iFrame resizing for all 3 LTI tools
- Implement scroll position caching for the gallery tool
- Improved sorting in engagement index
- Fixed bug in liking/disliking code
- Fixed removed activities bug in Points Configuration (https://jira.ets.berkeley.edu/jira/browse/CYB-212)
- UI code simplification where possible
- Consistently use Services, Factories and Controllers
- Added significant documentation throughout the UI codebase
- Added pointers for possible REST API improvements

As this is my first significant work using Angular, I would appreciate a thorough review that completely picks this apart ;) /cc @christianv @raydavis

### CODE REVIEW

File: app/assets/javascripts/angular/factories/galleryFactory.js [hunk 1]
```diff
@@ -3,42 +3,31 @@
...
-    var getSubmissions = function() {
-      // var url = '/dummy/json/gallerySubmissions.json';
-      var url = '/api/v1/gallery/index';
-      return $http.get(url);
+    /**
+     * Get all gallery items
+     *
+     * @return {Promise<GalleryItem[]>}      Promise returning all gallery items
+     */
+    var getGalleryItems = function() {
```
Comment (nicolaasmatthijs) [inline app/assets/javascripts/angular/factories/galleryFactory.js]: My preference is to use a consistent format and not unnecessarily condense things too much

File: app/assets/javascripts/angular/factories/galleryItemFactory.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Increment the number of views on a gallery item
+     *
+     * @param  {String}                 id            The id of the gallery item to increment the views for
+     * @return {Promise}                              $http promise
+     */
+    var incrementViews = function(id) {
+      var data = {
+        'gallery_id': id
```
Comment (nicolaasmatthijs) [inline app/assets/javascripts/angular/factories/galleryItemFactory.js]: Sure. What style guide are we trying to follow for our JS code?

File: app/assets/templates/gallery.html [hunk 3]
```diff
@@ -1,8 +1,8 @@
-<div class="dc-gallery dc-clearfix-container" ng-controller="GalleryController">
+<div class="dc-gallery dc-clearfix-container" data-ng-controller="GalleryController">
   <h1>
-    <a href="/canvas/embedded/gallery" data-ng-click="resetOptions()">Media Gallery</a>
+    <a href="/canvas/embedded/gallery" data-ng-click="resetSortAndFilter()">Media Gallery</a>
   </h1>
 
-  <div data-ng-if="!selectedItemId" data-ng-include="'gallery_list.html'"></div>
-  <div data-ng-if="selectedItemId" data-ng-include="'gallery_item.html'"></div>
+  <div data-ng-show="!selectedItemId" data-ng-include="'gallery_list.html'"></div>
```
Comment (nicolaasmatthijs) [inline app/assets/templates/gallery.html]: Actually, this is no longer necessary. I switched back to `ng-if`

**Comment by nicolaasmatthijs** [comment]
Followed up on review. Back to @christianv for final check and merging.

**Comment by christianv** [comment]
@nicolaasmatthijs lgtm

File: app/assets/javascripts/angular/factories/galleryItemFactory.js (hunk 2 above)
Comment (christianv) [inline app/assets/javascripts/angular/factories/galleryItemFactory.js]: @nicolaasmatthijs we have a very minimal styleguide at https://github.com/ets-berkeley-edu/calcentral/blob/master/docs/styleguide.md
but most of the rules are defined in the `.jscsrc` file:
https://github.com/ets-berkeley-edu/calcentral/blob/master/.jscsrc

The one in datacultures hasn't been updated in a while.

**Pull request merged**